### PR TITLE
malformed_sstable_exception: log error

### PR DESCRIPTION
**Pull request by bhalevy**
log an error meassage whenever a malformed_sstable_exception is generated to make them more visible so they won't go unnoticed.

For example, as seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/4318/artifact/logs-gating.release.2/1640145203194_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_simple_kill_streaming_node_while_bootstrapping/node3.log
```
WARN  2021-12-22 03:53:22,082 [shard 0] stream_session - [Stream #b5002180-62da-11ec-9d9b-4e2d269411e2] stream_transfer_task: Fail to send to 127.0.74.4:0: sstables::malformed_sstable_exception (missing index entry for summary index 45 (bound 0x60000106c8f0) in sstable /jenkins/workspace/scylla-master/next/scylla/.dtest/dtest-ek7636zz/test/node3/data/ks3/standard1-92df8f0062da11ecb1ba221c1df2a2ea/md-2-big-Index.db)
WARN  2021-12-22 03:53:22,082 [shard 0] stream_session - [Stream #b5002180-62da-11ec-9d9b-4e2d269411e2] Failed to send: sstables::malformed_sstable_exception (missing index entry for summary index 45 (bound 0x60000106c8f0) in sstable /jenkins/workspace/scylla-master/next/scylla/.dtest/dtest-ek7636zz/test/node3/data/ks3/standard1-92df8f0062da11ecb1ba221c1df2a2ea/md-2-big-Index.db)
```
    
The malformed_sstable_exception may be printed only at a WARN level or might be swallawed as a whole.

### CODE REVIEW

File: sstables/exceptions.hh [hunk 1]
```diff
@@ -22,23 +22,33 @@
 
 #pragma once
 
+#include <experimental/source_location>
```
Comment (denesb) [inline sstables/exceptions.hh:25]: Unused include.

File: sstables/exceptions.hh [hunk 2]
```diff
@@ -22,23 +22,33 @@
...
+
 #include <seastar/core/print.hh>
 
 #include "seastarx.hh"
+#include "log.hh"
```
Comment (denesb) [inline sstables/exceptions.hh:30]: Unneeded include, you can forward declare the logger.

File: sstables/sstables.cc [hunk 3]
```diff
@@ -824,15 +826,15 @@ future<> sstable::read_toc() noexcept {
...
+                throw_malformed_sstable_exception(sstlog, *this, "Empty TOC", filename(component_type::TOC));
             }
         });
     }).then_wrapped([this] (future<> f) {
         try {
             f.get();
         } catch (std::system_error& e) {
             if (e.code() == std::error_code(ENOENT, std::system_category())) {
-                throw malformed_sstable_exception(filename(component_type::TOC) + ": file not found");
+                throw_malformed_sstable_exception(sstlog, *this, format("{}", std::current_exception()), filename(component_type::TOC));
```
Comment (denesb) [inline sstables/sstables.cc:837]: Changing the log message should probably be done in its own commit.

File: sstables/sstables.cc [hunk 4]
```diff
@@ -40,6 +41,7 @@
 #include <seastar/coroutine/maybe_yield.hh>
 
 #include "dht/sharder.hh"
+#include "sstables/component_type.hh"
```
Comment (denesb) [inline sstables/sstables.cc:44]: Is this needed?

**Review by denesb — COMMENTED** [review]
LGTM, some minor comments.

This will probably result in most messages being logged twice. Don't know what we can do about it though.

**Comment by bhalevy** [comment]
In v2:
- remove unneeded includes
- add forward declaration of seastar::logger in sstables/exceptions.hh
- extracted patch changing the exception message when a component is not found.

File: sstables/exceptions.hh (hunk 1 above)
Comment (bhalevy) [reply sstables/exceptions.hh]: removed in v2

File: sstables/exceptions.hh (hunk 2 above)
Comment (bhalevy) [reply sstables/exceptions.hh]: removed in v2 and replaced with forward declaration

File: sstables/sstables.cc (hunk 3 above)
Comment (bhalevy) [reply sstables/sstables.cc:835]: done in v2

File: sstables/sstables.cc (hunk 4 above)
Comment (bhalevy) [reply sstables/sstables.cc]: removed in v2

File: sstables/sstables.cc [hunk 5]
```diff
@@ -3140,6 +3140,21 @@ std::optional<large_data_stats_entry> sstable::get_large_data_stat(large_data_ty
     return std::make_optional<large_data_stats_entry>();
 }
 
+malformed_sstable_exception::malformed_sstable_exception(const sstable& sst, sstring msg, const sstring& filename)
+    : malformed_sstable_exception(fmt::format("{} in sstable {} origin={}", msg, filename, sst.get_origin()))
+{
+}
+
+[[noreturn]] void throw_malformed_sstable_exception(logging::logger& log, const sstable& sst, sstring msg, const sstring& filename) {
```
Comment (denesb) [inline sstables/sstables.cc:3148]: With a fresh mind in a fresh year I realized that nothing stops new code from just using `throw malformed_sstable_exception` as before, completely circumventing this new feature. If covering all throw sites both present and future is your intention, this might be better served by making a logger a mandatory constructor argument of `malformed_sstable_exception` and doing the logging in the constructor. That way all future code will automatically benefit from this patch.

Comment (bhalevy) [reply sstables/sstables.cc:3148]: That would work, though I think that passing a logger to the exception constructor is quite ugly.
I can also rename the exception type to e.g. `_malformed_sstable_exception` so it would be clearer that it's an internal name.

Plus, there was at least one call site, in `sstable::read_simple` where I left the code unchanged, as follows:
```
        } catch (malformed_sstable_exception &e) {
            throw malformed_sstable_exception(e.what(), file_path);
        }
```

since the code was just wrapping a nested malformed_sstable_exception and adding the file_path, since lower-level functions that are static are not aware of the sstable file_path.

File: test/boost/broken_sstable_test.cc [hunk 6]
```diff
@@ -232,25 +232,25 @@ SEASTAR_TEST_CASE(empty_toc) {
 
 SEASTAR_TEST_CASE(alien_toc) {
     return broken_sst("test/resource/sstables/badtoc", 2,
-               "test/resource/sstables/badtoc/la-2-big-Statistics.db: file not found");
+               "open failed: No such file or directory [test/resource/sstables/badtoc/la-2-big-Statistics.db]");
```
Comment (avikivity) [inline test/boost/broken_sstable_test.cc:235]: This will fail if the error is `Nie ma takiego pliku ani katalogu`.

File: sstables/sstables.cc [hunk 7]
```diff
@@ -3140,6 +3140,21 @@ std::optional<large_data_stats_entry> sstable::get_large_data_stat(large_data_ty
...
+    auto ex = std::make_exception_ptr(malformed_sstable_exception(sst, msg, filename));
+    log.error("{} found at {}", ex, current_backtrace());
+    std::rethrow_exception(std::move(ex));
+}
```
Comment (avikivity) [inline sstables/sstables.cc:3152]: Won't this cause the exception to be double-printed? at the call site and at the catch site.

Also, it makes it impossible to expect the exception. For example, importing sstables might want to catch the exception and report it to the API, not log it (poor example, but it shows that the function presupposes that the exception cannot be handled).

**Comment by bhalevy** [comment]
This PR was absndoned.  Closing
